pkg/controller/sync/aggregator/aggregator.go

```diff
@@ -0,0 +1,105 @@
+package aggregator
+
+import (
+	"fmt"
+	gosync "sync"
+
+	"golang.org/x/exp/maps"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type KindName struct {
+	Kind string
+	Name string
+}
+
+func NewGVKAggregator(allowedKeyKinds []string) *GVKAgreggator {
+	return &GVKAgreggator{
+		store:        make(map[KindName]map[schema.GroupVersionKind]struct{}),
+		reverseStore: make(map[schema.GroupVersionKind]map[KindName]struct{}),
+
+		allowedKeyKinds: allowedKeyKinds,
+	}
+}
+
+// todo comments.
+type GVKAgreggator struct {
+	mu gosync.RWMutex
+
+	store        map[KindName]map[schema.GroupVersionKind]struct{}
+	reverseStore map[schema.GroupVersionKind]map[KindName]struct{}
+
+	allowedKeyKinds []string
+}
+
+func (b *GVKAgreggator) validateKind(k KindName) error {
+	if len(b.allowedKeyKinds) == 0 {
+		return nil
+	}
+
+	for _, kk := range b.allowedKeyKinds {
+		if k.Kind == kk {
+			// we found a match so allow the kind
+			return nil
+		}
+	}
+
+	return fmt.Errorf("unsupported key kind for key (%s, %s); supported values are: %s", k.Kind, k.Name, b.allowedKeyKinds)
+}
+
+func (b *GVKAgreggator) ListGVKs() []schema.GroupVersionKind {
```

Comment:
What downstream functionality is `ListGVKs` intended to support?

Reply:
Removed in favor of `IsPresent`

See https://github.com/open-policy-agent/gatekeeper/pull/2733#discussion_r1184363258 for more context